core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java

```diff
@@ -441,4 +447,70 @@ public void testDropNamespace() throws IOException {
     FileSystem fs = Util.getFs(new Path(metaLocation), conf);
     Assert.assertFalse(fs.isDirectory(new Path(metaLocation)));
   }
+
+  @Test
+  public void testVersionHintFile() throws Exception {
+    Configuration conf = new Configuration();
+    String warehousePath = temp.newFolder().getAbsolutePath();
+    HadoopCatalog catalog = new HadoopCatalog(conf, warehousePath);
+
+    // Create a test table with multiple versions
+    TableIdentifier tableId = TableIdentifier.of("tbl");
+    Table table = catalog.createTable(tableId, SCHEMA, PartitionSpec.unpartitioned());
+
+    DataFile dataFile1 = DataFiles.builder(SPEC)
+        .withPath("/a.parquet")
+        .withFileSizeInBytes(10)
+        .withRecordCount(1)
+        .build();
+
+    DataFile dataFile2 = DataFiles.builder(SPEC)
+        .withPath("/b.parquet")
+        .withFileSizeInBytes(10)
+        .withRecordCount(1)
+        .build();
+
+    table.newAppend().appendFile(dataFile1).commit();
+    table.newAppend().appendFile(dataFile2).commit();
+    long secondSnapshotId = table.currentSnapshot().snapshotId();
+
+    // Get the version-hint.text file location
+    String versionHintLocation = ((HadoopTableOperations) catalog.newTableOps(tableId)).versionHintFile().toString();
+
+    // Write old data to confirm that we are writing the correct file
+    FileIO io = new HadoopFileIO(conf);
+    io.deleteFile(versionHintLocation);
+    try (PositionOutputStream stream = io.newOutputFile(versionHintLocation).create()) {
+      stream.write("1".getBytes(StandardCharsets.UTF_8));
+    }
+
+    // Load the table and check the current snapshotId
+    Assert.assertEquals(secondSnapshotId, catalog.loadTable(tableId).currentSnapshot().snapshotId());
+
+    // Write newer data to confirm that we are writing the correct file
+    io.deleteFile(versionHintLocation);
+    try (PositionOutputStream stream = io.newOutputFile(versionHintLocation).create()) {
+      stream.write("3".getBytes(StandardCharsets.UTF_8));
+    }
+
+    // Load the table and check the current snapshotId
+    Assert.assertEquals(secondSnapshotId, catalog.loadTable(tableId).currentSnapshot().snapshotId());
+
+    // Write an empty version hint file
+    io.deleteFile(versionHintLocation);
+    io.newOutputFile(versionHintLocation).create().close();
+
+    // Load the table and check the current snapshotId
+    Assert.assertEquals(secondSnapshotId, catalog.loadTable(tableId).currentSnapshot().snapshotId());
+
+    // Just delete the file - double check that we have manipulated the correct file
+    io.deleteFile(versionHintLocation);
+
+    // Check that exception is thrown
+    AssertHelpers.assertThrows(
+        "Should not be able to find the table",
+        NoSuchTableException.class,
+        "Table does not exist: tbl",
+        () -> catalog.loadTable(tableId));
```

Comment:
I don't think this is correct. If the hint file is missing, the table should still be readable.

Reply:
Thanks for the review @rdblue!
That was another codepath which I did not changed to keep the changes minimal, but I agree with you comment, and corrected that path as well.